### PR TITLE
Bug 1486869 - Make Clear Private Data settings panel more granular

### PR DESCRIPTION
**Pull request by mrachamallu**
*Thank you for submitting a pull request, your contributions are greatly appreciated!*

Added a new option in settings to manage data for specific website.

## Pull Request Checklist

- [ ] My patch has gone through review and I have addressed review comments
- [ ] My patch has a standard commit message that looks like `Bug 12345678 - This fixes something something`

- [ ] I have updated the *Unit Tests* to cover new or changed functionality
- [ ] I have updated the *UI Tests* to cover new or changed functionality
- [ ] I have marked the bug with `[needsuplift]`
- [ ] I have made sure that localizable strings use `NSLocalizableString()`

## Screenshots

If your patch includes user interface changes that you would like to suggest or that you would like UX to look at, please include them here.

## Notes for testing this patch

If useful, please leave notes for QA, explaining what this patch changes and how it can be best tested and verified.

### CODE REVIEW

**Comment by farhanpatel** [comment]
This looks really good Meera! I'm still going through it but one thing I noticed 
- remove a website via the search
- press cancel
- the website still appears in the list 

It looks like the website data is not refreshed after leaving a search

**Comment by mrachamallu** [comment]
Thanks for the comments, Farhan
-I had implemented a function to reload data after pressing cancel on search and the websites I delete disappear from the list, not sure why it's not working for you, but I'll look into it again
-I changed WebsiteDataManagementViewController.swift to enum style. It looks a little different than ActivityStreamPanel.swift, but if it's all good, then I'll change the other two tableviews to this style as well

File: Client/Frontend/Settings/websiteSearchResultsViewController.swift [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+import Shared
+
+private let SectionHeaderFooterIdentifier = "SectionHeaderFooterIdentifier"
+
+class websiteSearchResultsViewController: UIViewController, UITableViewDataSource, UITableViewDelegate {
+
+    let flexible = UIBarButtonItem(barButtonSystemItem: UIBarButtonSystemItem.flexibleSpace, target: self, action: nil)
+    let editButton: UIBarButtonItem = UIBarButtonItem(title: "Edit", style: .plain, target: self, action: #selector(didPressEdit))
+    let doneButton: UIBarButtonItem = UIBarButtonItem(title: "Done", style: .done, target: self, action: #selector(didPressDone))
+    let deleteButton: UIBarButtonItem = UIBarButtonItem(title: "Delete", style: .plain, target: self, action: #selector(didPressDelete))
```
Comment (farhanpatel) [inline Client/Frontend/Settings/websiteSearchResultsViewController.swift]: These buttons need localized strings in Strings.swift

File: Client/Frontend/Settings/WebsiteDataManagementViewController.swift [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = UITableViewCell(style: .default, reuseIdentifier: nil)
+
+        let section = Section(rawValue: indexPath.section)!
+        switch section {
+        case .sites:
+            let site = siteRecords[indexPath.item]
+            cell.textLabel?.text = site.nameOfSite
+        case .showMore:
+            cell.textLabel?.text = "Show More"
```
Comment (farhanpatel) [inline Client/Frontend/Settings/WebsiteDataManagementViewController.swift]: Needs a localized string

File: Client/Frontend/Settings/WebsiteDataManagementViewController.swift [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+
+    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        let section = Section(rawValue: indexPath.section)!
+        switch section {
+        case .sites:
+            break
+        case .showMore:
+            getAllWebsiteData()
+        case .button:
+            func clearwebsitedata(_ action: UIAlertAction) {
```
Comment (farhanpatel) [inline Client/Frontend/Settings/WebsiteDataManagementViewController.swift]: Move the method outside of the `didSelectRowAt` it'll make it cleaner to understand

File: Client/Frontend/Settings/WebsiteDataManagementViewController.swift [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+                WKWebsiteDataStore.default().removeData(ofTypes: dataTypes, modifiedSince: .distantPast, completionHandler: { return })
+                siteRecords.removeAll()
+                showMoreButtonEnabled = false
+                tableView.reloadData()
+            }
+            let alert =  UIAlertController.clearWebsiteDataAlert(okayCallback: clearwebsitedata)
+            let generator = UIImpactFeedbackGenerator(style: .heavy)
```
Comment (farhanpatel) [inline Client/Frontend/Settings/WebsiteDataManagementViewController.swift]: Nice 😄

File: Client/Frontend/Settings/ClearPrivateDataTableViewController.swift [hunk 5]
```diff
@@ -100,76 +105,91 @@ class ClearPrivateDataTableViewController: ThemedTableViewController {
...
-            self.clearables
-                .enumerated()
-                .compactMap { (i, pair) in
-                    guard toggles[i] else {
-                        return nil
+        if indexPath.section == SectionArrow {
+            let view = WebsiteDataManagementViewController()
+            navigationController?.pushViewController(view, animated: true)
+        } else if indexPath.section == SectionButton {
+            func clearPrivateData(_ action: UIAlertAction) {
```
Comment (farhanpatel) [inline Client/Frontend/Settings/ClearPrivateDataTableViewController.swift]: Moving this method outside of `didSelectRowAt` will clean this up nicely

File: Client/Frontend/Settings/WebsiteDataManagementViewController.swift [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+
+        //Search Controller setup
+        let searchResults = websiteSearchResultsViewController()
+        searchController = UISearchController(searchResultsController: searchResults)
+        searchController.searchResultsUpdater = searchResults
+        searchController.obscuresBackgroundDuringPresentation = false
+        searchController.searchBar.placeholder = "Filter Sites"
+        searchController.searchBar.delegate = self
+        if #available(iOS 11.0, *) {
+            navigationItem.searchController = searchController
```
[the remark this answers is not in the document]
Comment (mrachamallu) [reply Client/Frontend/Settings/WebsiteDataManagementViewController.swift]: updated the else statement so searchController is added to the titleView

**Pull request merged**